File: transpiler/src/test/java/com/here/ivi/api/test/NiceErrorCollector.java
/*
 * Copyright (C) 2017 HERE Global B.V. and its affiliate(s). All rights reserved.
 *
 * This software, including documentation, is protected by copyright controlled by
 * HERE Global B.V. All rights are reserved. Copying, including reproducing, storing,
 * adapting or translating, any or all of this material requires the prior written
 * consent of HERE Global B.V. This material also contains confidential information,
 * which may not be disclosed to others without prior written consent of HERE Global B.V.
 *
 */

package com.here.ivi.api.test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

import java.util.concurrent.Callable;
import org.junit.rules.ErrorCollector;

/**
 * Custom error collector that calls specific assert*() methods instead of just generic assertThat()
 */
public class NiceErrorCollector extends ErrorCollector {

  /**
   * Adds a failure with the given {@code message} to the table if {@code actualValue} is null.
   * Execution continues, but the test will fail at the end if the match fails.
   */
  public <T> void checkNull(final String message, final T value) {
    checkSucceeds(
        (Callable<Object>)
            () -> {
              assertNull(message, value);
              return value;
            });
  }

  /**
   * Adds a failure with the given {@code message} to the table if {@code actualValue} is not null.
   * Execution continues, but the test will fail at the end if the match fails.
   */
  public <T> void checkNotNull(final String message, final T value) {
    checkSucceeds(
        (Callable<Object>)
            () -> {
              assertNotNull(message, value);
              return value;
            });
  }

  /**
   * Adds a failure with the given {@code message} to the table if {@code expectedValue} does not
   * match {@code actualValue}. Execution continues, but the test will fail at the end if the match
   * fails.
   *
   * @implNote Calling assertEquals() makes sure the stored AssertionError is actually a
   *     ComparisonFailure object. This allows IntelliJ to format the message nicely with a link to
   *     a diff of the values.
   */
  public <T> void checkEquals(final String message, final T expectedValue, final T actualValue) {
    checkSucceeds(
        (Callable<Object>)
            () -> {
              assertEquals(message, expectedValue, actualValue);
              return actualValue;
            });
  }
}

File: gluecodium/src/test/java/com/here/gluecodium/test/NiceErrorCollector.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.gluecodium.test

import org.bitbucket.cowwoc.diffmatchpatch.DiffMatchPatch
import org.hamcrest.CoreMatchers.containsString
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertThat
import org.junit.ComparisonFailure
import org.junit.rules.ErrorCollector

/**
 * Custom error collector that calls specific assert*() methods instead of just generic assertThat().
 */
internal class NiceErrorCollector : ErrorCollector() {
    /**
     * Adds a failure with the given [message] to the table if [value] is not null.
     * Execution continues, but the test will fail at the end if a match fails.
     */
    fun <T> checkNotNull(message: String, value: T?) {
        checkSucceeds { assertNotNull(message, value) }
    }

    /**
     * Adds a failure with the given [message] to the table if the [expectedValue] string is not contained in the
     * [actualValue] string. Execution continues, but the test will fail at the end if a match fails. The assertion
     * error is rethrown as a [ComparisonFailure] to produce a nice diff in IDE.
     */
    fun checkContainsString(message: String, expectedValue: String, actualValue: String) {
        checkSucceeds {
            try {
                assertThat(message, actualValue, containsString(expectedValue))
            } catch (e: AssertionError) {
                val strippedValue = stripValue(actualValue, expectedValue)
                throw ComparisonFailure(message, expectedValue, strippedValue)
            }
        }
    }

    /**
     * Strip prefix and suffix from the [actualValue] value to show only the diff to the [expectedValue] itself.
     */
    private fun stripValue(actualValue: String, expectedValue: String): String {
        var result = actualValue

        val diff = DiffMatchPatch().diffMain(expectedValue, actualValue)
        if (diff.first.operation == DiffMatchPatch.Operation.INSERT) {
            result = result.drop(diff.first.text.length)
        }
        if (diff.last.operation == DiffMatchPatch.Operation.INSERT) {
            result = result.dropLast(diff.last.text.length)
        }

        return result
    }
}
